### ISSUE TITLE
Exposing the inner ref of an element via BlazorComponent returns an emtpy(?) ref

### ISSUE THREAD

**Issue by chanan** [aspnet/Blazor#1756]
I have a component that inherits from BlazorComponent with this code:

```
protected override void BuildRenderTree(RenderTreeBuilder builder)
        {
            base.BuildRenderTree(builder);
            builder.OpenElement(0, TagName);
            foreach (var param in _attributesToRender)
            {
                builder.AddAttribute(1, param.Key, param.Value);
            }
            if (_classname != null) builder.AddAttribute(1, "class", _classname);
            builder.AddElementReferenceCapture(2, capturedRef => { ElementRef = capturedRef; });
            builder.AddContent(3, _childContent);
            builder.CloseElement();
        }
```

It's used like so:

```
<form onSubmit={this.alertValue}>
    <Dynamic TagName="input" ElementRef="@inputRef" />
        <button type="button" onclick="@OnClick">
            Submit
        </button>
</form>

@functions {
    ElementRef inputRef;

    async void OnClick()
    {
        await inputRef.AlertValue();
    }
}
```

Which ultimately prints out to the console what the ref is. What gets printed is:

```
{
  "_blazorElementRef": null
}
```

**Comment by enetstudio** [comment]
Mate, ElementRef is an opaque reference object holding a reference to an Html element, but it cannot be used in C#. You use it only when you want to pass a reference to an Html element to JavaScript only. The parameter accepted in a JavaScript object, such as a function, behaves just as if it was an object reference retrieved, say, by getElementById()

**Comment by enetstudio** [comment]
@chanan,  you can read about the ElementRef here: https://blazor.net/docs/javascript-interop.html

**Comment by enetstudio** [comment]
@chanan,

Once again, the ElementRef is not intended to be used in your C# code. It can and should only be passed to a JavaScript function. Now, if you want to pass an ElementRef to the user of your component, I don't see anything that could stop you from doing it. Just define a property of type ElementRef with get and set accessors, and pass it to whomever you want. But again, the receiver of that reference (meaning element reference) can do nothing with it, such as accessing the properties of the element. The only thing the receiver of your element reference can do is pass it to a JavaScript function...

**Comment by chanan** [comment]
Yes, i am aware of that and that is what I am doing. The output I showed on my bug is from a console.log in JavaScript.

**Comment by AndreasTruetschel** [comment]
This is the the value that is marshalled to javascript, if the `ElementRef` on the C# side does not point to a valid html element. I came accross the same issue when defining a javascript library where a non-present javascript reference was a valid case and I wanted to handle this case directly in JS. My solution to this is (in JS):

```
window.alertValue= function(htmlElement) {
   if(!htmlElement || !htmlElement._blazorElementRef) {
      // There was no html element reference passed. Handle this case accordingly.
      console.log("No element ref captured.");
   }
   else {
      console.log("element: %O", element);
   }
}
```

I think in your case, you have to use two way binding of the `ElementRef` property of your component. The `inputRef` varaible of your outer component is that updated, when the `Dynamic` component updates its `ElementRef` property currently.

**Comment by chanan** [comment]
@AndreasTruetschel Yes, I think that is what is happening. I am not sure how to fix it though.

I thought you might have meant changing the outer component to:
```
<Dynamic TagName="input" ElementRef="@inputRef" />
```
but that throws an error:
```
ncaught (in promise) Error: System.ArgumentException: 'bind' does not accept values of type Microsoft.AspNetCore.Blazor.ElementRef. To read and write this value type, wrap it in a property of type string with suitable getters and setters.
  at Microsoft.AspNetCore.Blazor.Components.BindMethods.SetValueHandler[T]
```
